lib/ui/painting/flutter_rtree.cc

```diff
@@ -0,0 +1,243 @@
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/flutter_rtree.h"
+#include "flutter/fml/logging.h"
+
+namespace flutter {
+
+FlutterRTree::FlutterRTree() : fCount(0) {}
+
+void FlutterRTree::insert(const SkRect boundsArray[],
+                          const SkBBoxHierarchy::Metadata metadata[],
+                          int N) {
+  SkASSERT(0 == fCount);
+
+  std::vector<Branch> branches;
+  branches.reserve(N);
+
+  for (int i = 0; i < N; i++) {
+    const SkRect& bounds = boundsArray[i];
+    if (bounds.isEmpty()) {
+      continue;
+    }
+    Branch b;
+    b.fBounds = bounds;
+    b.fOpIndex = i;
+    b.isDraw = (metadata == nullptr) ? false : metadata[i].isDraw;
+    branches.push_back(b);
+  }
+
+  fCount = (int)branches.size();
+  if (fCount) {
+    if (1 == fCount) {
+      fNodes.reserve(1);
+      Node* n = this->allocateNodeAtLevel(0);
+      n->fNumChildren = 1;
+      n->fChildren[0] = branches[0];
+      fRoot.fSubtree = n;
+      fRoot.fBounds = branches[0].fBounds;
+    } else {
+      fNodes.reserve(CountNodes(fCount));
+      fRoot = this->bulkLoad(&branches);
+    }
+  }
+}
+
+void FlutterRTree::insert(const SkRect boundsArray[], int N) {
+  // TODO(egarciad): Ask Skia about removing this method signature.
+  insert(boundsArray, nullptr, N);
+}
+
+FlutterRTree::Node* FlutterRTree::allocateNodeAtLevel(uint16_t level) {
+  SkDEBUGCODE(Node* p = fNodes.data());
+  fNodes.push_back(Node{});
+  Node& out = fNodes.back();
+  SkASSERT(fNodes.data() == p);  // If this fails, we didn't reserve() enough.
+  out.fNumChildren = 0;
+  out.fLevel = level;
+  return &out;
+}
+
+// This function parallels bulkLoad, but just counts how many nodes bulkLoad
+// would allocate.
+int FlutterRTree::CountNodes(int branches) {
+  if (branches == 1) {
+    return 1;
+  }
+  int numBranches = branches / kMaxChildren;
+  int remainder = branches % kMaxChildren;
+  if (remainder > 0) {
+    numBranches++;
+    if (remainder >= kMinChildren) {
+      remainder = 0;
+    } else {
+      remainder = kMinChildren - remainder;
+    }
+  }
+  int currentBranch = 0;
+  int nodes = 0;
+  while (currentBranch < branches) {
+    int incrementBy = kMaxChildren;
+    if (remainder != 0) {
+      if (remainder <= kMaxChildren - kMinChildren) {
+        incrementBy -= remainder;
+        remainder = 0;
+      } else {
+        incrementBy = kMinChildren;
+        remainder -= kMaxChildren - kMinChildren;
+      }
+    }
+    nodes++;
+    currentBranch++;
+    for (int k = 1; k < incrementBy && currentBranch < branches; ++k) {
+      currentBranch++;
+    }
+  }
+  return nodes + CountNodes(nodes);
+}
+
+FlutterRTree::Branch FlutterRTree::bulkLoad(std::vector<Branch>* branches,
+                                            int level) {
+  if (branches->size() == 1) {  // Only one branch.  It will be the root.
+    return (*branches)[0];
+  }
+
+  // We might sort our branches here, but we expect Blink gives us a reasonable
```

Comment:
This comment is irrelevant. Does this condition hold for us? Either way, please update the comment.